src/apps/pfu.cpp

```diff
@@ -0,0 +1,195 @@
+// got changed in the code without retrospectively adjusting the player's stats.
+void ResetModifiableAttrs::execute(Char &ch) const {
+    OBJ_DATA *obj;
+    // Restore the character to his/her true condition
+    // All of the attributes reset here are ones that can be modified in AFFECT_DATA.cpp
+    ranges::fill(ch.mod_stat, 0);
+    ch.max_hit = ch.pcdata->perm_hit;
+    ch.max_mana = ch.pcdata->perm_mana;
+    ch.max_move = ch.pcdata->perm_move;
+    // #216 -1 is initial value for armour now
+    ranges::fill(ch.armor, -1);
+    ch.hitroll = 0;
+    ch.damroll = 0;
+    ch.saving_throw = 0;
+    ch.sex = ch.pcdata->true_sex;
+    // add back object effects
+    for (auto loc = 0; loc < MAX_WEAR; loc++) {
+        obj = get_eq_char(&ch, loc);
+        if (!obj)
+            continue;
+        for (size_t i = 0; i < ch.armor.size(); i++)
+            ch.armor[i] -= apply_ac(obj, loc, i);
+
+        if (!obj->enchanted)
+            for (const auto &af : obj->pIndexData->affected)
+                af.apply(ch);
+
+        for (auto &af : obj->affected)
+            af.apply(ch);
+    }
+    // add back spell effects
+    for (auto &af : ch.affected)
+        af.apply(ch);
+}
+
+CharUpgraderResult::CharUpgraderResult() : ch_(nullptr), upgraded_(false) {}
+
+CharUpgraderResult::CharUpgraderResult(Char *ch, bool upgraded) : ch_(ch), upgraded_(upgraded) {}
+
+CharUpgraderResult::~CharUpgraderResult() {
+    if (ch_) {
+        extract_char(ch_, true);
+    }
+}
+
+CharUpgraderResult::operator bool() const noexcept { return ch_ && !ch_->name.empty(); }
+
+CharUpgraderResult::operator Char &() const noexcept { return *ch_; }
+
+CharUpgrader::CharUpgrader(std::string_view name, const Tasks &all_tasks, Logger &logger)
+    : name_(name), desc_(Descriptor(0)), all_tasks_(all_tasks), logger_(logger) {}
+
+auto CharUpgrader::required_tasks(const Char &ch) const {
+    return all_tasks_ | ranges::views::filter([&ch](const auto &upgrade) { return upgrade->is_required(ch); });
+}
+
+CharUpgraderResult CharUpgrader::upgrade() {
+    if (auto ch = simulate_login()) {
+        bool upgraded{};
+        logger_.debug("{}", name_);
+        // Chars that have an malformed pfile will have been created and added
+        // to the world, but no tasks should be run. Released when the result is destroyed.
+        if (!ch->name.empty()) {
+            for (auto &task : required_tasks(*ch)) {
+                task->execute(*ch);
+                upgraded = true;
+            }
+        } else
+            logger_.warn("Invalid player file: {}", name_);
+        return {ch, upgraded};
+    }
+    return {};
+}
+
+Char *CharUpgrader::simulate_login() {
+    if (auto fp = WrappedFd::open(filename_for_player(name_))) {
+        LastLoginInfo last_login;
+        // This snippet is very similar to what's in nanny() and try_load_player().
+        Char *ch = new Char();
```

Comment:
(non-actionable) hopefully soon we can remove these "scary" naked `new`s. But for now we're definitely stuck with them :)